Address Ben Schwartz' IANA expert review on -15 (#48)
core-wg/draft-dns-over-coap: draft-ietf-core-dns-over-coap.md

The terms "payload" and "body" are used as defined in {{-coap-blockwise, Section 2}}.
Note that, when block-wise transfer is not used, the terms "payload" and "body" are to be understood as equal.

> For better readability, examples in this document the binary payload and resource records are provided in hexadecimal representation as well as a human-readable format.

Reviewer: this sentence appears still broken. it should be "examples in this document of the binary payload and resource records are provided [...]"

{{-core-dnr}} provides  further exploration of the challenges here.

This document specifies "docpath" as
a single-valued SvcParamKey that is mandatory for DoC SVCB records.
If the "docpath" SvcParamKey is absent, the service should not be considered a valid DoC service.

The presentation format value of "docpath" MUST be the absolute path in the URI path-abempty form
(see {{-uri, Section 3.3}}).

The wire-format value for "docpath" consists of 0 or more sequences of octets prefixed by their
respective length as a single octet, the length octet, and these length-value pairs are concatenated
to form the SvcParamValue.
These pairs MUST exactly fill the SvcParamValue; otherwise, the SvcParamValue is malformed.
Each such length-value pair represents one segment of the absolute path to the DoC resource.
0 length-value pairs represent the root path "/".
Note that this restricts the length of each segment to at most 255 octets.
Paths with longer segments can not be advertised with the "docpath" SvcParam and are thus NOT
RECOMMENDED for the path to the DoC resource.

> Note that this wire-format uses the same encoding as the "alpn" SvcParam.

Reviewer 2: That makes sense for consistency. After all, that representation happens in a DNS server that is likely to have *that* implemented, and used by an operator who is familiar with DNS conventions, and constrained devices never see that anyway.

{{-core-dnr}} provides  further exploration of the challenges here.

This document specifies "docpath" as
a single-valued SvcParamKey that is mandatory for DoC SVCB records.
If the "docpath" SvcParamKey is absent, the service should not be considered a valid DoC service.

The docpath is devided up into segments of the absolute path to the DoC resource (docpath-segment),
each a sequence of 1-255 octets.
In ABNF {{-abnf}}:

~~~
docpath-segment = 1*255OCTET
~~~

Note that this restricts the length of each docpath-segment to at most 255 octets.

> Paths with longer segments can not be advertised with the "docpath" SvcParam and are thus NOT

Reviewer 3: ```suggestion
Paths with longer segments cannot be advertised with the "docpath" SvcParam and are thus NOT
```

RECOMMENDED for the path to the DoC resource.

The presentation format value of "docpath" SHALL be a comma-sperated list ({{Appendix A.1 of -svcb}})
of 0 or more docpath-segments.
The root path "/" is represented by 0 docpath-segments, i.e., an empty list.
The same considerations for the "," and "\" characters in docpath-segments for zone-file
implementations as for the alpn-ids in an "alpn" SvcParam MAY apply ({{Section 7.1.1 of -svcb}}).

The wire-format value for "docpath" consists of 0 or more sequences of octets prefixed by their
respective length as a single octet, the length octet, and these length-value pairs are concatenated
to form the SvcParamValue.
These pairs MUST exactly fill the SvcParamValue; otherwise, the SvcParamValue is malformed.
Each such length-value pair represents one segment of the absolute path to the DoC resource.
The root path "/" is represented by 0 length-value pairs, i.e., SvcParamValue length 0.

> Note, that this format uses the same encoding as the "alpn" SvcParam and can just reuse the

Reviewer 3: ```suggestion
Note that this format uses the same encoding as the "alpn" SvcParam and can reuse the
```

> decoders and encoders for that SvcParam with the adaption that 0 elements are allowed.

Reviewer 3: ```suggestion
decoders and encoders for that SvcParam with the adaption that a length of zero is allowed.
```

> As long as each docpath-segment are of length 0 and 24 octets, it is easily transferred into the path

Reviewer 3: ```suggestion
As long as each docpath-segment is of length 0 and 24 octets, it is easily transferred into the path
```

{{-core-dnr}} provides  further exploration of the challenges here.

This document specifies "docpath" as
a single-valued SvcParamKey that is mandatory for DoC SVCB records.
If the "docpath" SvcParamKey is absent, the service should not be considered a valid DoC service.

> The value of "docpath" MUST be a CBOR sequence of 0 or more text strings (see

Reviewer 4: I recommending reusing the binary format from "dohpath" or "alpn" if possible.  It seems likely that either of these formats will suffice, and that one or both of them is already needed for both these clients and DNS servers.  These formats are no less compact than the proposed CBOR format, and could be implemented without importing a CBOR library in the authoritative DNS server.

A dependency on CBOR is likely to present a considerable barrier to implementation for authoritative DNS servers, which do not generally include CBOR functionality today.  Servers could implement a special-purpose serializer for this tiny subset of CBOR, but this presents no obvious benefit over the other syntax options.

Author: > I recommending reusing the binary format from "dohpath" or "alpn" if possible. It seems likely that either of these formats will suffice, and that one or both of them is already needed for both these clients and DNS servers. These formats are no less compact than the proposed CBOR format, and could be implemented without importing a CBOR library in the authoritative DNS server.

The idea was that constrained using DoC would have a CBOR parser anyway, while on a large server would suffer less from an additional parser. Additionally, this makes the path compatible with representation in CRIs (https://datatracker.ietf.org/doc/draft-ietf-core-href/).

However, having gone through the examples now and revisiting the encoding for "alpn", I think indeed that it has the same benefits of a CBOR parser without putting the dependency burden on the DNS servers. The conversion to a CRI path would be just as simple as it is currently to convert from CBOR to, e.g., the Uri-Path option.

Author: That is, if I understand correctly and the equivalent of an "alpn"-like encoding, e.g., for "/n/s" would be (taking 0a for now to make things less confusing by a new value):

    00 0a 00 04 01 6e 01 73

Author: Adopted in 3fe26f035d927c2f227321e3183f992015c3a5e0

A more generalized construction algorithm for any CoAP request can be found in {{-transport-indication}}.

### Examples

A typical SVCB resource record response for a DoC server at the root path "/" of the server looks

> like the following (the "docpath" SvcParam is the last 3 bytes "0a 00 00" in the binary):

Reviewer 2: ```suggestion
like the following (the "docpath" SvcParam is the last 4 bytes "00 0a 00 00" in the binary):
```

I'm no expert, but IIRC it's two 16-bit values, and that'd be consistent with the `00 01 00 03` value `02 63 6f` of a single 2-byte ALPN "co".

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 06 28 00 1e 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 00

    Resource record (human-readable):
      _dns.example.org.  1576  IN SVCB 1 dns.example.org (
          alpn=co docpath= )

Though the root path is RECOMMENDED, here are two examples with the "docpath" as paths of varying
depth for illustration purposes, first "/dns" (the last 8 bytes "00 0a 00 04 63 64 6e 73", "63"
marking the CBOR element as a text string of length 3, see {{Section 3 of -cbor}}):

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 00 55 00 22 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 04 63 64 6e 73

    Resource record (human-readable):
      _dns.example.org.    85  IN SVCB 1 dns.example.org (
          alpn=co docpath="dns" )

> Here an examples for the path "/n/s" (the last 8 bytes "00 0a 00 04 61 6e 61 73", "61" marking each

Reviewer: Second, [...]

CBOR element as a text string of length 1, see {{Section 3 of -cbor}})):

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 06 6b 00 22 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 04 61 6e 61 73

    Resource record (human-readable):
      _dns.example.org.   643  IN SVCB 1 dns.example.org (
          alpn=co docpath="n","s" )

If the server also provides DNS over HTTPS, "dohpath" and "docpath" MAY co-exist, though it is

> RECOMMENDED to use a separate SVCB record for each service:

Reviewer 4: If you think it's generally operationally preferable for these endpoints to have distinct priorities, that seems fine to note here.  Personally, I don't know of a reason why distinct priorities would typically be necessary.

Note that (per RFC 9461) each protocol can use its default port number while sharing a single record.

like the following (the "docpath" SvcParam is the last 4 bytes "00 0a 00 00" in the binary):

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 06 28 00 1e 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 00

    Resource record (human-readable):
      _dns.example.org.  1576  IN SVCB 1 dns.example.org (
          alpn=co docpath )

The root path is RECOMMENDED but not required. Here are two examples where the "docpath" represents
paths of varying depth. First, "/dns" is provided in the following example
(the last 8 bytes "00 0a 00 04 63 64 6e 73", "63" marking the CBOR element as a text string of
length 3, see {{Section 3 of -cbor}}):

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 00 55 00 22 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 04 63 64 6e 73

    Resource record (human-readable):
      _dns.example.org.    85  IN SVCB 1 dns.example.org (
          alpn=co docpath="dns" )

> Second, an examples for the path "/n/s" (the last 8 bytes "00 0a 00 04 61 6e 61 73", "61" marking

Reviewer 4: It looks like this presumes a value of 0x000a for the "docpath" registration.  Ideally I think the example would use a value from the reserved range (65280+) until a final value is assigned by IANA.

Also, I believe "61" should be "63" in this example, as it is a string of length 3.

each CBOR element as a text string of length 1, see {{Section 3 of -cbor}})):

    Resource record (binary):
      04 5f 64 6e 73 07 65 78 61 6d 70 6c 65 03 6f 72
      67 00 00 40 00 01 00 00 06 6b 00 22 00 01 03 64
      6e 73 07 65 78 61 6d 70 6c 65 03 6f 72 67 00 00
      01 00 03 02 63 6f 00 0a 00 04 61 6e 61 73

    Resource record (human-readable):
      _dns.example.org.   643  IN SVCB 1 dns.example.org (

>           alpn=co docpath="n","s" )

Reviewer 4: Yes, I was speaking only about the textual representation.

>           alpn=co docpath="n","s" )

Author: I used the path-style syntax now in 3fe26f035d927c2f227321e3183f992015c3a5e0
